lib/location/method_wifi.c

```diff
@@ -0,0 +1,307 @@
+/*
+ * Copyright (c) 2021 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <stdio.h>
+#include <zephyr.h>
+#include <assert.h>
+
+#include <logging/log.h>
+
+#include <net/net_if.h>
+#include <net/net_event.h>
+#include <net/wifi_mgmt.h>
+
+#include <modem/location.h>
+
+#include "loc_core.h"
+#include "loc_utils.h"
+
+#include "rest/rest_services_wifi.h"
+
+LOG_MODULE_DECLARE(location, CONFIG_LOCATION_LOG_LEVEL);
+
+BUILD_ASSERT(
+	IS_ENABLED(CONFIG_LOCATION_METHOD_WIFI_SERVICE_NRF_CLOUD) ||
+	IS_ENABLED(CONFIG_LOCATION_METHOD_WIFI_SERVICE_HERE) ||
+	IS_ENABLED(CONFIG_LOCATION_METHOD_WIFI_SERVICE_SKYHOOK),
+	"At least one WiFi positioning service must be enabled");
+
+extern location_event_handler_t event_handler;
+extern struct loc_event_data current_event_data;
+
+struct k_work_args {
+	struct k_work work_item;
+	struct loc_wifi_config wifi_config;
+	int64_t starting_uptime_ms;
+
+};
+
+static struct net_if *wifi_iface;
+
+static struct k_work_args method_wifi_start_work;
+
+static bool running;
+
+static uint32_t current_scan_result_count;
+static uint32_t latest_scan_result_count;
+
+struct method_wifi_scan_result {
+	char mac_addr_str[WIFI_MAC_ADDR_STR_LEN + 1];
+	char ssid_str[WIFI_SSID_MAX_LEN + 1];
+	uint8_t channel;
+	int8_t rssi;
+};
+
+static struct method_wifi_scan_result
+	latest_scan_results[CONFIG_LOCATION_METHOD_WIFI_SCANNING_RESULTS_MAX_CNT];
+static K_SEM_DEFINE(wifi_scanning_ready, 0, 1);
+
+/******************************************************************************/
+
+static int method_wifi_scanning_start(void)
+{
+	int ret;
+
+	LOG_DBG("Triggering start of WiFi scanning");
+
+	latest_scan_result_count = 0;
+	current_scan_result_count = 0;
+
+	if (wifi_iface != NULL) {
+		ret = net_mgmt(NET_REQUEST_WIFI_SCAN, wifi_iface, NULL, 0);
+		if (ret) {
+			LOG_ERR("Failed to initiate WiFi scanning: %d", ret);
+			ret = -1;
+		}
+	} else {
+		LOG_ERR("Failed to initiate WiFi scanning: no wifi device");
+		ret = -1;
+	}
+	return ret;
+}
+
+/******************************************************************************/
+
+static void method_wifi_handle_wifi_scan_result(struct net_mgmt_event_callback *cb)
```

Comment:
I'm thinking about the double "wifi" in the function name (same also with next function) and whether we could just use `method_wifi_handle_scan_result`. I guess it comes from the event name `NET_EVENT_WIFI_SCAN_RESULT`. We could also have `method_wifi_scan_result_handle`, which would align better to the style of having the verb at the end. Not sure what's the best but somehow this double "wifi" bothers me :)